src/wasm-binary.h

```diff
@@ -898,6 +918,18 @@ class WasmBinaryWriter : public Visitor<WasmBinaryWriter, void> {
   void visitBlock(Block *curr) {
     if (debug) std::cerr << "zz node: Block" << std::endl;
     o << int8_t(BinaryConsts::Block);
+
+    int arity = curr->type != none && curr->type != unreachable;
+    if (brTargetArities.count(curr->name)) {
+      if (curr->type == unreachable) {
+        arity = brTargetArities[curr->name];
+      } else {
+        assert((curr->type != none) == brTargetArities[curr->name]);
+      }
+    }
+    // For blocks with type unreachable but whose breaks have arity 1, encode i32 as their
+    // signature so that the decoder knows to pop a value for the breaks' values.
+    o << binaryWasmType(curr->type != unreachable ? curr->type : arity ? i32 : none);
```

Comment:
Once we parse block signatures, we can just encode `curr->type` here which should never be unreachable. Likewise for ifs and loops.